Add context menu in prompt dialog (git cola --prompt)

nakanoi opened the pull request:
According to this issue, "[Cannot remove old folders from git-cola main window #1199](https://github.com/git-cola/git-cola/issues/1199)", I implemented context menu in prompt dialog (`git cola --prompt`).

The content of the context menu is same as favorite(and recent) widget (in `cola/widgets/bookmarks.py`) in main UI.
A lot of context menu functions come from the code.

- Open
- Open in New WIndow
- Open Using Default Application
- Copy
- Launch Editor
- Launch Terminal
- Clear Deafult Repository
- Rename Repository
  - **only activated in "Folder" tab** [1]
- Delete from Bookmakrs List
  - **only in `Favorites` view**
- Delete from Recent List
  - **only in `Recent` view**
- Prune Missing Entries

### NOTE
> [1]
> In "List" tab, each item text is its repository's full path like `/home/foo/some-repository`.
> So, in "List" tab, set all items UNEDITABLE.

![prompt-capture](https://user-images.githubusercontent.com/72122101/210164192-aa14e82e-4852-4336-ac95-a572d72c0997.png)

Fixes #1199 
**Thank you!**

davvid commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+                ):
+                    other_items.remove(other_item)
+                    other_model.removeRow(other_index.row())
+                    return
```
This is pretty complex because it's trying to mutate the view to quickly remove this entry.

The original code was a little fast and loose in that it modified the models directly but it wasn't as complex. Would it be possible to eliminate this coupling?

This code also originally only dealt with the bookmarks list. It'd almost be better to not change that behavior for now so that it doesn't need to have this complicated "other model" coupling going on.

Update: after testing, I see a little bit of why this is being done -- the view is now split and it uses a different widget for the Recents and the Favorites.

I'll have to think about that.. in the past it was all presented as a single widget that contained both bookmarks and recent items. The rationale is that a single widget is easier to use and understand, even if it is a little overloaded under the hood.

With two widgets it means there are two things to scroll and there's some redundancy across the lists because repos can be in both the Recents and Favorites, which is not quite as nice.

What do you think? Is it better to present them in a single view or is the screen real estate and complexity of two views worth it?

nakanoi replied:
I see. I had no such idea.
I splited into two views because of the reasons below. But doing that made many things complecated and might feels redundant.

> - View spliting
> I thought too.
> But a case like below could be possible. `bookmarks` and `recent` widgets in main UI display repository's name, and I thought splitiing the view is one of the solutions for clearfying which name should be displayed.
> And I also thought that displaying two views seems to be more usable.
> But after I read all comments, I though *view shouldn't be splited*.

So, changing the code to display only bookmarks list view might be good. Is it OK?

davvid commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+            model = self.recent_model
+            self.recent_items = new_items = []
+        else:
+            return
```
Is the `else:` ever reached, or is it unreachable?

This is another example where the internal complexity might be able to be reduced. The `mode` is basically deciding which of the internal two models to use. That seems like it can be improved by factoring that stuff out.

nakanoi replied:
This `else:` never reached and for completeness too, and it might be redundant.
OK, I'll eliminate it.

davvid commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+        self.setIcon(icon)
+        self.setText(name)
+        self.setToolTip(path)
+        self.setFlags(self.flags() | Qt.ItemIsEditable)
```
Is it better to use `setEditable(bool)` rather than using `setFlags()`?

davvid replied:
Also.. it looks like we call `setEditable(False)` right above here...

nakanoi replied:
Oh, using `setEditable(bool)` is enough and `setFlags` is overload.
I'm really sorry.

davvid commented on cola/widgets/startup.py:
```diff
@@ -68,46 +76,52 @@ def __init__(self, context, parent=None):
-        self.short_paths = short_paths
+        self.recent_title_label = qtutils.label(
+            text="Recent", align=Qt.AlignHCenter | Qt.AlignVCenter, selectable=False
+        )
```
TBD whether splitting Recents and Favorites is worth it, but just a note that all translatable UI strings should use the `N_(...)` function for i18n.

This should be `N_('Recents')` here, and `N_('Favorites')` down below.

nakanoi replied:
OK. I'll change it.

davvid commented on cola/widgets/startup.py:
```diff
@@ -7,18 +7,29 @@
+LIST_VIEW_BOOKMARK = 0
+LIST_VIEW_RECENT = 1
+
+LABEL_TITLE_FONT_SIZE = 16
```
We probably don't want to hard-code a font size... I'll see where it's used to suggest an alternative.

nakanoi replied:
Yes, as this comment, I'll improve it.

>Instead of using a hard-coded font size, lookup the current font size and then scale by some factor (eg. 0.8). Is the default size too small or too large?

davvid commented on cola/widgets/startup.py:
```diff
@@ -68,46 +76,52 @@ def __init__(self, context, parent=None):
+        )
+        title_font = self.recent_title_label.font()
+        title_font.setBold(True)
+        title_font.setPixelSize(LABEL_TITLE_FONT_SIZE)
```
Instead of using a hard-coded font size, lookup the current font size and then scale by some factor (eg. `0.8`). Is the default size too small or too large?

nakanoi replied:
That is the coolest way and I've never come up with it.
Thank you! I'll improve it.

davvid commented on cola/widgets/startup.py:
```diff
@@ -158,24 +167,58 @@ def __init__(self, context, parent=None):
+            for item in items:
+                if isinstance(item, PromptWidgetItem):
+                    item_name = item.change_path_text(name_mode)
+                    model.rename_entry(item, item_name)
```
I'm confused about this part -- is this going and calling `rename_entry()` on each item?

In theory that's renaming all of the repos. Am I reading that correctly?

I recall that switching between modes was pretty slow.. if so, this might explain why. It's actually renaming things to their same values and saving the settings file on each line in this loop.

That seems.. not good. That makes me think we shouldn't even allow renaming in this view. What do you think?

nakanoi replied:
Yes, it is corrent. And I think so too.

When change Tabs between "Folder" and "List", item text should be changed(`name` <=> `path`).
I added `rename_entry()` here because somehow `item.setText()` in `change_path_text()` emits `itemChnaged` signal. So, `context.settings.bookmarks` and `context.settings.recent` get changed.

To keep consistency, call `rename_entry()` after `change_path_text()` call.
But, by disabling renaming, `itemChange` signal becomes unnecessarry and this part gets to be able to eliminated.

davvid reviewed (changes requested):
Thanks for picking this up. I've got a bunch of question and notes, and there's a key behavior in `tab_changed()` that looks like a bug.

I think it'd be good to add more context menu actions, but I would suggest that we not add renaming, and that we try to avoid splitting the view if possible.

I'm open to hearing other opinions as to whether or not we should split the views, but that's my current take on it at this time.

Let me know what you think.

davvid commented:
Meta: we have this weird special `Open...` folder entry that behaves unlike the rest of them in the UI.

I almost think we should remove that special entry and rename the `Open...` button on the bottom left to `Browse...` for clarity. I suspect that might make the code a little simpler as well.

nakanoi commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+                paths.add(normalize(bookmark['path']))
+        elif self._is_recent_mode():
+            for recent in settings.recent:
+                paths.add(normalize(recent['path']))
```
I see. Yes, `_is_bookmarks_mode()` and `_is_recents_mode()` become unnecessary.
I'll try to change it!

nakanoi commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+            for recent in settings.recent:
+                paths.add(normalize(recent['path']))
+
+        if isinstance(item, PromptWidgetItem):
```
This is because each view("bookmarks" and "recent") has `Open...` item, and it is `QStandardItem`, not `PromptWidgetItem`. Not to open context menu with `Open...` item click, I added `isinstance` filter.
 
> Meta: we have this weird special Open... folder entry that behaves unlike the rest of them in the UI.
> 
> I almost think we should remove that special entry and rename the Open... button on the bottom left to Browse... for clarity. I suspect that might make the code a little simpler as well.

As following this comment, removing `Open...` item and all `isinstance(item, PromptWidgetItem)` filters from code seems good. How about this?

nakanoi commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+    def remove_missing(self):
+        """Remove missing entries from the favorites/recent file list"""
+        settings = self.context.settings
+        settings.remove_missing_recent()
```
OK! I'll do it.

davvid replied:
About `remove_missing` -- I think this action doesn't even need to be item-specific. I think we would want to remove missing bookmarks and missing recent entries all in a single call here.

For what it's worth, the remove-missing context menu entry seems like it should be present even when right-clicking in the widget in the blank area without any icons. Not important for now, but just noting in case it helps. This should be the only UI action that will modify both the bookmarks and recents from a single action. The rest can limit their scope to bookmarks when `is_bookmark` is True.

nakanoi commented on cola/widgets/startup.py:
```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+
+
+def _disable_rename(_x, _y, _z):
+    return False
```
this method is never reached and for completeness.
OK, I'll remove it.

nakanoi commented on cola/widgets/startup.py:
```diff
@@ -158,24 +167,58 @@ def __init__(self, context, parent=None):
+            (self.recent, self.recent_items),
+        ]
+        for model, items in models_items:
+            if idx == 0:
```
Yes! I'll improve it.

nakanoi commented:
Thank you for detailed comments!

- Disable rename entry action
I agree to it. I added this action because `bookmarks` and `recent` widgets have it. So I'm going to remove it from context menu.

- View spliting
I thought too.
But a case like below could be possible. `bookmarks` and `recent` widgets in main UI display repository's name, and I thought splitiing the view is one of the solutions for clearfying which name should be displayed.
And I also thought that displaying two views seems to be more usable.
But after I read all comments, I though *view shouldn't be splited*.

```python
# Double name entry for the same repository("path": "/foo")
context.settings.bookmarks = [
  {"name": "foo", "path": "/foo"},
]
context.settings.recent = [
  {"name": "buzz", "path": "/foo"},
]
```

> This code also originally only dealt with the bookmarks list. It'd almost be better to not change that behavior for now so that it doesn't need to have this complicated "other model" coupling going on.

So, changing the code to display only bookmarks list view might be good. How about it?

---
I'm very motivated to improve all of them. Thank you for your cooperation!

davvid commented:
> Meta: we have this weird special `Open...` folder entry that behaves unlike the rest of them in the UI.
> 
> I almost think we should remove that special entry and rename the `Open...` button on the bottom left to `Browse...` for clarity. I suspect that might make the code a little simpler as well.

We can ignore this comment now. I familiarized myself with the code and made some minor aesthetic changes to the UI to make that item a little more intuitive. Please rebase.

I've renamed the special `Open...` item to `Browse...`. We'll keep that entry. It's convenient because it lets us hit enter when the dialog pops up to go straight into a file browser.

The separate button, which also had an `Open...` label, was renamed to `Open Git Repository` so that it makes it a little clearer that pressing it will open the currently selected item.

I want to call out this section in the original code where it builds the items:

```python
        bookmarks = settings.bookmarks
        recent = settings.recent
        all_repos = bookmarks + recent
        ...
        for repo in all_repos:
            ...
            item = QtGui.QStandardItem(path)
```

If we stick to a single view then we can push the "is this path bookmarked?" knowledge down into the item at this point by tweaking this loop. eg. by using `enumerate` we can detect whether we're looping over a bookmark or a recent repo.

```python
for idx, repo in enumerate(all_repos):
    is_bookmark = idx < len(bookmarks)
    ...
```

and then it can use `is_bookmark` when building the item so that the context menu can be customized?

I take it back about not having the "Rename" action -- that's pretty useful so no reason not to have if possible.

Sorry for not being in favor of having two list views instead of one. If anything, I hope it might simplify the implementation a bit.

Also, it might be the case that we want to remove a bookmark but still keep the item in the "Recents" area, so having the handling be that it detects bookmarks and only removes Bookmarks/Favorites when detected, and otherwise attempts to remove from the Recents is probably the right behavior. That's better than removing from both lists using a single action.

I hope that makes sense.

Also, not sure if you noticed, but the code uses the term "bookmark" but the UI presents it as "Favorites". That's a historical thing -- it was originally called "Bookmarks" to match firefox jargon. #289 is when that was renamed. We didn't change the class names when we did that -- we only changed the labels use the term "Favorites" but kept everything else using the Bookmarks nomenclature.

Let me know what you think so far. Here's a quick summary of my general sugs...

* Use a custom item so that it can remember whether or not it's from the bookmarks
* If it's from the Bookmarks then the bookmark-related remove/rename should be used
* If it's from the Recents then the recents-related remove/rename should be used
* Keep all of the context menu actions you've added. Those are great.

Thanks again!

davvid commented:
>  But after I read all comments, I though view shouldn't be splited.

Thank you, I kinda reached the same conclusion. I was a little on the fence too..

it's kinda nice with them split, but sometimes less is more, so I tend to go with a "simpler" choice when possible. It also helps out new users that don't yet have any Recents of Favorites ~ seeing the `Browse...` entry in two lists without anything else in them might be kinda confusing. Thanks again!

davvid commented:
(I think this is implied in what I wrote earlier, but I didn't respond to this question so I'll do so now)

> So, changing the code to display only bookmarks list view might be good. Is it OK?

We'll want to continue displaying both the Bookmarks and the Recent items, in the same order it currently displays them (bookmarks first, and then recent items) in a single View. Using the `is_bookmark` flag to special-case the Items seems like it should give us enough context to specialize their context menu entries and editable states.

Having some `if self.is_bookmark` logic to decide on the functions/cmds to run down inside the item hopefully won't be too bad. Better might be to push that into BuildItem so that the custom Item class can be simple, but it's fine if it's just a small handful of spot checks.

nakanoi commented:
Thank you for many great tips and idea!

It's OK. Implementing SIngle View might make things simpler.
I got understand  that "bookmarks" and "favorite". I'll care of them.
I'll add `is_bookmark` flag attribute to custom Item Class(`PromptWidgetItem`) for identifying between bookmarks and recent.

I think your suggestions are best and make sense. I'll try to implement improvemnt all of them!

I might ask some questions while implemnting them, please forgive me.
Thank you!

nakanoi commented:
Hi @davvid!
I implemented all given suggestions!

- Changed to single view
- Removed followings
  - `_disable_rename()` and unreachable `else` statements
  - unnecessary `setFlags()` and remain `setEditable()`
  - translatable `N_('Favorites')` and `N_('Recent')`
    - There is no need to use them because labels become unnecessary.
  - Font size hard coding
    - Same as `N_('Favorites')` one.
- `remove_missing()`
  - Remove missing repository from both bookmarks and recent list.
- Use custom item to keep whether the item comes from bookmarks list or not
- Delete/Rename item....
  - item in bookmarks list if it comes from bookmarks list
  - item in recent list if it comes from recent list
  - item in bookmarks list if it comes from both bookmarks and recent list

---

On the other hand,

- Still use `isinstance(item, PromptWidgetItem)`
  - To filter items because `self.items` contains a `QStandardItem` for `Browse...` item and repository items for `PromptWidgetItem`.
- Remove `model.rename_entry()` call.
If `item.setText()` called, then `itemChanged` signal will be emitted and that slot changes item name unproperly.
So, instead of calling `item.setText()`, I implemented `tab_change` to re-instancinate all items like `StartupDialog.set_model()`.

---

If I misunderstood something, please tell me.
Thank you!

davvid reviewed (changes requested):
This looks really great, it's almost ready to merge. Mostly small cosmetic things at this point.

Once it's ready, please rebase this branch and squash the commits together a bit. We won't really need the intermediate commits -- might as well squash things together a bit more. It seems like 2 or 3 commits at most should be sufficient:

* `starup.py: Change item to QStandardItem implement`
* [new squashed commit that adds the new widgets and tests]

If you're comfortable with git you can even tease apart a 3rd commit so that the tests can be committed separately.

Thanks for taking it this far! It's pretty much ready to go.. this last part is mostly just patch polish. This looks great!

nakanoi commented:
Thank you for re-reviewing!
After change, I'll clearfy commits and push it!

nakanoi commented:
Hi @davvid!
I fixed code based on the gaven comment and reabase commits!
Thank you.

davvid commented:
👏 👏 👏  very nicely done! Thanks again for improving cola.

nakanoi commented:
@davvid
Thank you for merging this PR!
I could'nt be done without your greateful help. 
Thank you so much again!
